src/emulator/auth/operations.ts

```diff
@@ -230,6 +231,157 @@ function lookup(
+    const userInfo = reqBody.users[index];
+
+    try {
+      assert(userInfo.localId, "localId is missing");
+      const uploadTime = new Date();
+      const fields: Omit<Partial<UserInfo>, "localId"> = {
+        displayName: userInfo.displayName,
+        photoUrl: userInfo.photoUrl,
+        lastLoginAt: userInfo.lastLoginAt,
+      };
+
+      // password
+      if (userInfo.passwordHash) {
+        // TODO: Check and block non-emulator hashes.
+        fields.passwordHash = userInfo.passwordHash;
+        fields.salt = userInfo.salt;
+        fields.passwordUpdatedAt = uploadTime.getTime();
+      } else if (userInfo.rawPassword) {
+        fields.salt = userInfo.salt || "fakeSalt" + randomId(20);
+        fields.passwordHash = hashPassword(userInfo.rawPassword, fields.salt);
+        fields.passwordUpdatedAt = uploadTime.getTime();
+      }
+
+      // custom attrs
+      if (userInfo.customAttributes) {
+        validateSerializedCustomClaims(userInfo.customAttributes);
+        fields.customAttributes = userInfo.customAttributes;
+      }
+
+      // federated
+      if (userInfo.providerUserInfo) {
+        fields.providerUserInfo = [];
+        for (const providerUserInfo of userInfo.providerUserInfo) {
+          const { providerId, rawId } = providerUserInfo;
+          // These providers are handled automatically by create / update.
+          if (providerId !== PROVIDER_PASSWORD && providerId !== PROVIDER_PHONE) {
+            assert(
+              providerId && rawId,
+              "((Auth Emulator only supports providerUserInfo with providerId AND rawId))"
+            );
+            const existingUserWithProviderRawId = state.getUserByProviderRawId(providerId, rawId);
+            assert(
+              !existingUserWithProviderRawId,
+              `((Auth Emulator does not support importing duplicate Provider id(${providerId}), Raw id(${rawId})))`
+            );
+            // TODO: Populate providerId based on federatedId.
+            fields.providerUserInfo.push({ ...providerUserInfo, providerId, rawId });
+          }
+        }
+      }
+
+      // phone number
+      if (userInfo.phoneNumber) {
+        assert(isValidPhoneNumber(userInfo.phoneNumber), "phone number format is invalid");
+        fields.phoneNumber = userInfo.phoneNumber;
+      }
+      // TODO: Support MFA.
+
+      fields.validSince = toUnixTimestamp(uploadTime).toString();
+      fields.createdAt = uploadTime.toString();
+      if (fields.createdAt && !isNaN(Number(userInfo.createdAt))) {
+        fields.createdAt = userInfo.createdAt;
+      }
+      if (userInfo.email) {
+        const email = userInfo.email;
+        assert(isValidEmailAddress(email), "email is invalid");
+        const existingUserWithEmail = state.getUserByEmail(email);
+        assert(
+          !existingUserWithEmail || existingUserWithEmail.localId === userInfo.localId,
+          `((Auth Emulator does not support importing duplicate email: ${email}))`
```

Comment:
TODO: Production catches this case only if `sanityCheck && emailRequired`. We should use production error message in that case, otherwise throw this emulator specific error message (since we do not want to support this edge case)